db/change_cache.go

```diff
@@ -184,29 +183,34 @@ func (c *changeCache) Init(dbcontext *DatabaseContext, notifyChange func(base.Se
 
 	heap.Init(&c.pendingLogs)
 
-	// background tasks that perform housekeeping duties on the cache
-	bgt, err := NewBackgroundTask("InsertPendingEntries", c.context.Name, c.InsertPendingEntries, c.options.CachePendingSeqMaxWait/2, c.terminator)
-	if err != nil {
-		return err
-	}
-	c.backgroundTasks = append(c.backgroundTasks, bgt)
-
-	bgt, err = NewBackgroundTask("CleanSkippedSequenceQueue", c.context.Name, c.CleanSkippedSequenceQueue, c.options.CacheSkippedSeqMaxWait/2, c.terminator)
-	if err != nil {
-		return err
-	}
-	c.backgroundTasks = append(c.backgroundTasks, bgt)
+	go func() {
+		dcpStartedErr := <-dcpStarted
+		// TODO: how to test this error condition?
+		if dcpStartedErr != nil {
+			// error is handled in StartGOCB2DCPFeed
+			return
+		}
+		// background tasks that perform housekeeping duties on the cache
+		bgt, err := NewBackgroundTask("InsertPendingEntries", c.context.Name, c.InsertPendingEntries, c.options.CachePendingSeqMaxWait/2, c.terminator)
+		// TODO: how to test this error condition?
+		if err != nil {
+			base.ErrorfCtx(context.TODO(), "Failed to initialize task %w", err)
+			dbcontext.Close()
+		}
+		c.backgroundTasks = append(c.backgroundTasks, bgt)
 
-	// Lock the cache -- not usable until .Start() called.  This fixes the DCP startup race condition documented in SG #3558.
-	c.lock.Lock()
+		bgt, err = NewBackgroundTask("CleanSkippedSequenceQueue", c.context.Name, c.CleanSkippedSequenceQueue, c.options.CacheSkippedSeqMaxWait/2, c.terminator)
+		// TODO: how to test this error condition?
+		if err != nil {
+			base.ErrorfCtx(context.TODO(), "Failed to initialize task %w", err)
+			dbcontext.Close()
+		}
+		c.backgroundTasks = append(c.backgroundTasks, bgt)
+	}()
 	return nil
 }
 
 func (c *changeCache) Start(initialSequence uint64) error {
-
-	// Unlock the cache after this function returns.
-	defer c.lock.Unlock()
-
 	// Set initial sequence for sequence buffering
 	c._setInitialSequence(initialSequence)
```

Comment:
The initial sequence is still set and read from places that are acquiring the lock (`DocChanged`, `changeCache.Clear`, `processEntry`, ...)

I'm not convinced this lock is safe to remove given the description of #3558

Reply:
Added back locking, since it will trigger the race detector in a test.